### PR TITLE
server: make the Drain interfaces available to SQL-only servers

### LINKED ISSUE #74412: server: equip the `mt-start-sql` server code with a drain process
Epic CRDB-11087

This is a pre-req to solving #72045

To add a feature to "execute some SQL statements before a node starts draining" (#72045), we must have a draining process to start with. The `start` and `start-single-node` commands provide that, but `mt start-sql` doesn't.

**Describe the solution you'd like**

We want to provide a Drain RPC and a shutdown sequence to the `mt start-sql` server process that is using a common code path with the one used for regular DB nodes.

### PR DESCRIPTION
**Pull request by knz**
Fixes #74412

First commit from #76292; the reviewers are invited to ignore it in this PR.

See individual commits for details.

cc @cockroachdb/obs-inf-prs

### CODE REVIEW

**Review by cameronnunez — APPROVED** [review]
:lgtm: Good work!

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 1 of 1 files at r3, 3 of 3 files at r4, 1 of 1 files at r5, 3 of 3 files at r6, 4 of 4 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76294#-:-MvZSDuz0_KvNPH5K9lk:b-pv0wtg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cameronnunez and @knz)

___
*[pkg/server/tenant.go, line 78 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76294#-MvZShFJALgM74QMnzck:-MvZShFJALgM74QMnzcl:b-97xtx0) ([raw file](https://github.com/cockroachdb/cockroach/blob/20bdaf146680441e81c2b09c98d1c5c0ec124c51/pkg/server/tenant.go#L78)):*
> ```Go
> }
> 
> // SQLServerWrapper is a utility structs that encapsulates
> ```

nit: "is a utility struct"

File: pkg/server/tenant_admin.go [hunk 1]
```diff
@@ -84,4 +99,52 @@ func (t *tenantAdminServer) Health(
...
+// instructs the process to terminate.
+// This method is part of the serverpb.AdminClient interface.
+func (t *tenantAdminServer) Drain(
+	req *serverpb.DrainRequest, stream serverpb.Admin_DrainServer,
+) error {
+	ctx := stream.Context()
+	ctx = t.AnnotateCtx(ctx)
+
+	// Which node is this request for?
+	parsedInstanceID, local, err := t.status.parseInstanceID(req.NodeId)
```
Comment (JeffSwenson) [inline pkg/server/tenant_admin.go:126]: FYI: We do not track instance ID in the CC operator. The operator thinks in terms of pods. Determining the instance ID of the pod we want to remove is more difficult than sending an RPC to the specific pod we want to drain. I suspect we will pass NodeId == "" when sending a Drain rpc.

**Review by jeffswenson — APPROVED** [review]
LGTM

It would be nice if there were an rpc to cancel a drain. There is a mismatch between the CRDB draining state and the serverless draining state. When the operator decides to reduce the number of pods allocated to a serverless cluster, it marks the surplus pods in a draining state. The sqlproxy maintains existing connections to draining pods, but it will not send new connections to them. If the cluster's utilization increases, the pods are put back in a serving state. If there was a way to cancel CRDB drains, we could notify the sql server it is draining as soon as the operator decides to reduce capacity. With one way drains, the operator needs to wait to send the drain until it is ready to remove the pod.

**Review by knz — COMMENTED** [review]
> It would be nice if there were an rpc to cancel a drain.

Point taken. let's discuss this in #76423.
I also encourage you to raise that issue to your nearest PM, so we can put it in our roadmap.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76294#-:-MvcfB2J-AxQSANkKfBw:bkcugrg)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cameronnunez and @JeffSwenson)

___
*[pkg/server/tenant.go, line 78 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76294#-MvZShFJALgM74QMnzck:-MvcfqVM97IHegx_n2ah:b-5098c6) ([raw file](https://github.com/cockroachdb/cockroach/blob/20bdaf146680441e81c2b09c98d1c5c0ec124c51/pkg/server/tenant.go#L78)):*
<details><summary><i>Previously, cameronnunez (Cameron Nunez) wrote…</i></summary><blockquote>

nit: "is a utility struct"
</blockquote></details>

Fixed. Thanks
___
*[pkg/server/tenant_admin.go, line 126 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76294#-MvU-3J4Vz2y4whodeqN-r6-126:-MvcfrX62QlHpg6TATGw:bnh6ueh) ([raw file](https://github.com/cockroachdb/cockroach/blob/cec0757bf9bbcd6ebd64c06a76b9b6dbda93f7b0/pkg/server/tenant_admin.go#L126)):*
> I suspect we will pass NodeId == "" when sending a Drain rpc.

Yes, that works.

The semantics as implemented here ensure that we're transparently compatible with what is supported on storage nodes, so the command-line tooling can stay the same.

**Comment by knz** [comment]
TFYRs!

bors r=cameronnunez,JeffSwenson

**Comment by knz** [comment]
bors r=cameronnunez,JeffSwenson

**Comment by knz** [comment]
hm lint failures
bors r-

**Comment by knz** [comment]
bors r=cameronnunez,JeffSwenson

**Pull request merged**